contracts/interchain-token-service/src/state.rs

```diff
@@ -159,6 +199,116 @@ pub fn unfreeze_chain(
         .change_context(Error::ChainNotFound(chain.to_owned()))
 }
 
+pub fn save_token_chain_info(
+    storage: &mut dyn Storage,
+    chain: ChainNameRaw,
+    token_id: TokenId,
+    token_chain_info: &TokenChainInfo,
+) -> Result<(), Error> {
+    TOKEN_CHAIN_INFO
+        .save(storage, &(chain, token_id), token_chain_info)
+        .change_context(Error::Storage)
+}
+
+pub fn may_load_token_chain_info(
+    storage: &dyn Storage,
+    chain: ChainNameRaw,
+    token_id: TokenId,
+) -> Result<Option<TokenChainInfo>, Error> {
+    TOKEN_CHAIN_INFO
+        .may_load(storage, &(chain, token_id))
+        .change_context(Error::Storage)
+}
+
+pub fn may_load_token_config(
+    storage: &dyn Storage,
+    token_id: &TokenId,
+) -> Result<Option<TokenConfig>, Error> {
+    TOKEN_CONFIGS
+        .may_load(storage, token_id)
+        .change_context(Error::Storage)
+}
+
+pub fn save_token_config(
+    storage: &mut dyn Storage,
+    token_id: &TokenId,
+    token_config: &TokenConfig,
+) -> Result<(), Error> {
+    TOKEN_CONFIGS
+        .save(storage, token_id, token_config)
+        .change_context(Error::Storage)
+}
+
+impl TokenSupply {
+    fn checked_add(&self, amount: nonempty::Uint256) -> Result<Self, OverflowError> {
+        match self {
+            TokenSupply::Tracked(supply) => {
+                TokenSupply::Tracked(supply.checked_add(amount.into())?)
+            }
+            TokenSupply::Untracked => TokenSupply::Untracked,
+        }
+        .then(Ok)
+    }
+
+    fn checked_sub(&self, amount: nonempty::Uint256) -> Result<Self, OverflowError> {
+        match self {
+            TokenSupply::Tracked(supply) => {
+                TokenSupply::Tracked(supply.checked_sub(amount.into())?)
+            }
+            TokenSupply::Untracked => TokenSupply::Untracked,
+        }
+        .then(Ok)
+    }
+}
```

Comment:
I don't think this should be implemented on the supply, this can lead to unexpected coding errors when `Untracked` can also add and sub

Reply:
In my mind, the caller shouldn't have to worry whether the supply underthehood is being tracked or not. But not a strong opinion, we could handle it more like an `Option`